classes/class-woothemes-sensei-certificates.php

```diff
@@ -1589,6 +1594,44 @@ public function add_certificate_button_to_current_course_completed_page() {
 		);
 	}
 
+	/**
+	 * Add "View Certificate" button to the Course List block patterns. To be
+	 * used with the sensei_course_list_block_patterns filter.
+	 *
+	 * @since 2.3.1
+	 *
+	 * @param array $patterns The block patterns.
+	 * @return array
+	 */
+	public function add_view_certificate_link_to_block_patterns( $patterns ) {
+		foreach ( $patterns as $key => $pattern ) {
+			// Only alter the templates for the Course List block, and only if
+			// it contains the Course Overview block.
+			if (
+				! in_array( 'query', $pattern['categories'], true )
+				|| ! in_array( 'core/query', $pattern['blockTypes'], true )
+				|| ! has_block( 'sensei-lms/course-overview', $pattern['content'] )
+			) {
+				continue;
+			}
+
+			// Add the block to the template. Note that we currently only
+			// support adding it after a "void" Course Overview block with no
+			// attributes. If, in the future, the Course Overview block can have
+			// inner content and/or attributes, then we will have to do a bit
+			// more advanced parsing here.
+			$content_with_block = preg_replace(
```

Comment:
I just want to highlight the fragility of this implementation. This will stop working completely if:

- The Course Overview link is removed from the template.
- Any attributes or inner content is added to the Course Overview link within the template.

I seriously considered addressing the second point by using either a more complex regex, or actually using a combination of `parse_blocks` and `serialize_blocks` to add our block in more robustly. In either case, it was going to be complex and seemed like overkill for this feature, especially for something that may never be an issue in practice. Plus, we don't have unit tests in this plugin, so I was worried about increasing the complexity too much.

That said, I am a bit concerned that the template will change in the future, and the "View Certificate" link will just quietly disappear. With this implementation, we would be unlikely to notice until a user notifies us. But I'm not sure if there's a better solution?

(If we ever pull Certificates into core, then this becomes a non-issue, because we can just add the block directly to the template, without needing a filter at all)

Reply:
Would it be simple if instead of the filter `sensei_course_list_block_patterns`, we create something like `sensei_course_list_block_patterns_extra_links`? It would be already in the middle of the string. WDYT?

Reply:
Hmm yeah maybe that would be better. I guess I was thinking that this solution would be the most flexible, but having a specific "slot" where we can add things seems tidier. I'll give that a try 👍

Reply:
Changed in https://github.com/woocommerce/sensei-certificates/pull/327/commits/69757e4d96dd27676bc1a4579ead541a7065bd71. I do like this better. Not as flexible, but I think it's a lot safer. Thanks fro the suggestion!

(Also changed in https://github.com/Automattic/sensei/pull/6296/commits/cf790563afcf1408ae1fbb79b2d13d850e87579c)